src/Servers/Kestrel/Core/src/Middleware/TlsListenerMiddleware.cs

```diff
@@ -41,20 +41,19 @@ internal async Task OnTlsClientHelloAsync(ConnectionContext connection)
             // no data is consumed, it will be processed by the follow-up middlewares
             input.AdvanceTo(buffer.Start);
 
-            switch (parseState)
+            if (parseState == ClientHelloParseState.NotEnoughData)
             {
-                case ClientHelloParseState.NotEnoughData:
-                    continue;
-
-                case ClientHelloParseState.NotTlsClientHello:
-                    await _next(connection);
-                    return;
+                continue;
+            }
 
-                case ClientHelloParseState.ValidTlsClientHello:
-                    _tlsClientHelloBytesCallback(connection, clientHelloBytes);
-                    await _next(connection);
-                    return;
+            if (parseState == ClientHelloParseState.ValidTlsClientHello)
+            {
+                _tlsClientHelloBytesCallback(connection, clientHelloBytes);
             }
+
+            // Here either it's a valid TLS client hello or definitely not a TLS client hello.
+            // Anyway we can continue with the middleware pipeline
```

Comment:
I'd prefer a 
```suggestion
            Debug.Assert(parseState is ValidTlsClientHello.NotTlsClientHello or ValidTlsClientHello.ValidTlsClientHello);
```
here, as it's kind of self-documenting, and if a new member gets added to the enum, a test will fail (i.e. the assert will trigger).

Note: you have to add the `System.Diagnostics` namespace too.

Reply:
yep, makes sense